### ISSUE TITLE
OGBG Pytorch deadlock with eval?

### ISSUE THREAD

**Issue by rakshithvasudev** [mlcommons/algorithmic-efficiency#149]
```
torchrun --no_python  --nproc_per_node 8 bash -c "python3 submission_runner.py --workload=ogbg --framework=pytorch --submission_path=reference_submissions/ogbg/ogbg_pytorch/submission.py --tuning_ruleset=external --tuning_search_space=refer
ence_submissions/ogbg/tuning_search_space.json --num_tuning_trials=1"
```

Looks like Ogbg Pytorch workload is not running? Not sure what I'm missing. It appears to be in a deadlock 

```
## Description
I0930 21:14:15.094216 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 1
I0930 21:14:15.113576 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 6
I0930 21:14:15.130159 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 5
I0930 21:14:15.146167 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 0
I0930 21:14:15.171331 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 3
I0930 21:14:15.186478 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 4
I0930 21:14:15.193458 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 7
I0930 21:14:15.208100 140737350199104 distributed_c10d.py:230] Added key: store_based_barrier_key:1 to store for rank: 2
I0930 21:14:15.208443 140737350199104 distributed_c10d.py:264] Rank 2: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.211157 140737350199104 distributed_c10d.py:264] Rank 5: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.211992 140737350199104 distributed_c10d.py:264] Rank 3: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.213966 140737350199104 distributed_c10d.py:264] Rank 7: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.214718 140737350199104 distributed_c10d.py:264] Rank 6: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.215536 140737350199104 distributed_c10d.py:264] Rank 1: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.217079 140737350199104 distributed_c10d.py:264] Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.217327 140737350199104 distributed_c10d.py:264] Rank 4: Completed store-based barrier for key:store_based_barrier_key:1 with 8 nodes.
I0930 21:14:15.645985 140737350199104 submission_runner.py:368] --- Tuning run 1/1 ---
I0930 21:14:15.646330 140737350199104 submission_runner.py:202] Initializing dataset.
I0930 21:14:15.646411 140737350199104 submission_runner.py:209] Initializing model.
I0930 21:14:31.463784 140737350199104 submission_runner.py:212] Initializing optimizer.
I0930 21:14:31.464552 140737350199104 submission_runner.py:224] Initializing metrics bundle.
I0930 21:14:31.464625 140737350199104 submission_runner.py:238] Starting training loop.
I0930 21:14:31.497103 140737350199104 xla_bridge.py:356] Unable to initialize backend 'tpu_driver': NOT_FOUND: Unable to find driver in registry given worker:
I0930 21:14:34.963119 140737350199104 xla_bridge.py:356] Unable to initialize backend 'rocm': NOT_FOUND: Could not find registered platform with name: "rocm". Available platform names are: Interpreter CUDA Host
I0930 21:14:34.963603 140737350199104 xla_bridge.py:356] Unable to initialize backend 'tpu': module 'jaxlib.xla_extension' has no attribute 'get_tpu_client'
2022-09-30 21:14:35.721456: W tensorflow/core/platform/cloud/google_auth_provider.cc:184] All attempts to get a Google authentication bearer token failed, returning an empty token. Retrieving token from files failed with "NOT_FOUND: Could not locate the credentials file.". Retrieving token from GCE failed with "FAILED_PRECONDITION: Error executing an HTTP request: libcurl code 6 meaning 'Couldn't resolve host name', error details: Could not resolve host: metadata".
I0930 21:14:35.997618 140737350199104 dataset_info.py:580] Load pre-computed DatasetInfo (eg: splits, num examples,...) from GCS: ogbg_molpcba/0.1.3
I0930 21:14:36.582294 140737350199104 dataset_info.py:491] Load dataset info from /tmp/tmpxpmqdojwtfds
I0930 21:14:36.585984 140737350199104 dataset_info.py:550] Field info.release_notes from disk and from code do not match. Keeping the one from code.
I0930 21:14:36.586203 140737350199104 dataset_info.py:550] Field info.splits from disk and from code do not match. Keeping the one from code.
I0930 21:14:36.586579 140737350199104 dataset_builder.py:437] Generating dataset ogbg_molpcba (/root/tensorflow_datasets/ogbg_molpcba/0.1.3)
Downloading and preparing dataset 37.70 MiB (download: 37.70 MiB, generated: 822.53 MiB, total: 860.23 MiB) to /root/tensorflow_datasets/ogbg_molpcba/0.1.3...
Dl Completed...: 0 url [00:00, ? url/s]          I0930 21:14:36.836208 140737350199104 download_manager.py:353] Downloading https://snap.stanford.edu/ogb/data/graphproppred/csv_mol_download/pcba.zip into /root/tensorflow_datasets/downloads/snap.stan.edu_ogb_grap_csv_mol_down_pcbapc4I82Cv1THcU-IggPHK8IHZ8qM-BJ3VDk-q_rtqrf4.zip.tmp.67c44807104745c4b9caa0a7f56bfff9...
Extraction completed...: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.46s/ file]
Dl Size...: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 37/37 [00:02<00:00, 15.02 MiB/s]
Dl Completed...: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.46s/ url]
Generating splits...:   0%|                                                                                                                                                                                                                                          | 0/3 [00:00<?, ? splits/sI0930 21:16:04.070159 140737350199104 writer.py:333] Done writing /root/tensorflow_datasets/ogbg_molpcba/0.1.3.incompleteD6PATR/ogbg_molpcba-train.tfrecord*. Number of examples: 350343 (shards: [43793, 43793, 43793, 43793, 43792, 43793, 43793, 43793])
Generating splits...:  33%|███████████████████████████████████████████████████████████████████████████▎                                                                                                                                                      | 1/3 [01:14<02:29, 74.70s/ splitsI0930 21:16:13.614588 140737350199104 writer.py:333] Done writing /root/tensorflow_datasets/ogbg_molpcba/0.1.3.incompleteD6PATR/ogbg_molpcba-validation.tfrecord*. Number of examples: 43793 (shards: [43793])
Generating splits...:  67%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋                                                                           | 2/3 [01:24<00:36, 36.34s/ splitsI0930 21:16:23.053446 140737350199104 writer.py:333] Done writing /root/tensorflow_datasets/ogbg_molpcba/0.1.3.incompleteD6PATR/ogbg_molpcba-test.tfrecord*. Number of examples: 43793 (shards: [43793])
Dataset ogbg_molpcba downloaded and prepared to /root/tensorflow_datasets/ogbg_molpcba/0.1.3. Subsequent calls will reuse this data.
I0930 21:16:23.071735 140737350199104 logging_logger.py:44] Constructing tf.data.Dataset ogbg_molpcba for split train, from /root/tensorflow_datasets/ogbg_molpcba/0.1.3
2022-09-30 21:16:23.077538: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0930 21:16:42.757913 140737350199104 spec.py:258] Evaluating on the training split.
I0930 21:16:42.761173 140737350199104 dataset_info.py:491] Load dataset info from /root/tensorflow_datasets/ogbg_molpcba/0.1.3
I0930 21:16:42.763895 140737350199104 dataset_builder.py:383] Reusing dataset ogbg_molpcba (/root/tensorflow_datasets/ogbg_molpcba/0.1.3)
I0930 21:16:42.764009 140737350199104 logging_logger.py:44] Constructing tf.data.Dataset ogbg_molpcba for split train[:10000], from /root/tensorflow_datasets/ogbg_molpcba/0.1.3
E0930 21:16:46.094066 140737350199104 submission_runner.py:307] Eval step 1 error.
Traceback (most recent call last):
  File "/work/algorithmic-efficiency/algorithmic_efficiency/workloads/ogbg/ogbg_pytorch/workload.py", line 68, in build_input_queue
    batch = next(dataset_iter)  # pylint: disable=stop-iteration-return
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "submission_runner.py", line 288, in train_once
    latest_eval_result = workload.eval_model(global_batch_size,
  File "/work/algorithmic-efficiency/algorithmic_efficiency/spec.py", line 259, in eval_model
    train_metrics = self._eval_model_on_split(
  File "/work/algorithmic-efficiency/algorithmic_efficiency/workloads/ogbg/workload.py", line 148, in _eval_model_on_split
    batch = next(self._eval_iters[split])
RuntimeError: generator raised StopIteration
```

![image](https://user-images.githubusercontent.com/13400803/193358157-1a0cc189-d97b-445c-8be2-d0d129e02c44.png)

**Comment by rakshithvasudev** [comment]
I'll close this for now, it might be my env issue

**Comment by rakshithvasudev** [comment]
I can see this happen with the specified versions of the libs in the Dockerfile

**Comment by rakshithvasudev** [comment]
I'm on the latest version of the submission runner (commit: eb8c53d726a1d098ea8d5a4b436282b277a175cc)

**Comment by runame** [comment]
Thanks for reporting the issue. At first glance it looks like you got a `StopIteration` error and no deadlock? I just confirmed that the workload still runs without issues on my setup.

**Comment by rakshithvasudev** [comment]
Yeah it just freezes there with the stop iteration perhaps leading to deadlock. 

I say deadlock because after the stopiteration, it doesn't return the control back to the OS. It just freezes there.

**Comment by znado** [comment]
I was also unable to reproduce this on our GCP VMs, maybe there is some issue with your env? Still very odd though. Can you run the jax version without issue?

**Comment by znado** [comment]
Going to close until someone can repro it in our GCP environment. We should eventually test that the code works in different environments, but we can do that more rigorously later once we know it works in ours.

**Comment by rakshithvasudev** [comment]
Sorry for the late reply. Been catching up with deadlines.

Thanks for closing. Yes! Indeed glad that it is not having any issues :)

**Comment by runame** [comment]
Ok, I think I was not using the latest code when trying to reproduce this. It is now indeed broken on `main`, the `StopIteration` in the first process is the issue. Will take a look.

**Comment by runame** [comment]
Will be fixed when PR #176 is merged (see PR description for the cause of the error).
